### PR TITLE
grafanaPlugins: fix update script/add some interesting data source plugins

### PR DESCRIPTION
**Pull request by nagisa**
## Description of changes

This adds a bunch of additional data source plugins for grafana. Most of these are grafana-maintained. I only added sources I myself am interested in.

One thing that came to mind is that this isn't something that needs to be manually maintained though. We could easily scrape the entire https://grafana.com/grafana/plugins/ on each update.nix invocation if we wanted to, I… think.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/servers/monitoring/grafana/plugins/grafana-googlesheets-datasource/default.nix [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{ grafanaPlugin, lib }:
+
+grafanaPlugin rec {
+  pname = "grafana-googlesheets-datasource";
+  version = "1.2.5";
+  zipHash = "sha256-AzH1rZFqEH8sovZZfJykvsEmCedEZWigQFHWHl6/PdE=";
```
Comment (nagisa) [inline pkgs/servers/monitoring/grafana/plugins/grafana-googlesheets-datasource/default.nix:6]: This zipHash repeats a few times and I can't figure out how to get the update script to fetch the right thing. It looks like `nix-prefetch-url` fails with a HTTP 400 for these, while it works fine if I just curl.

File: pkgs/common-updater/scripts/update-source-version [hunk 2]
```diff
@@ -165,7 +165,7 @@ else
     die "Couldn't figure out where out where to patch in new version in '$attr'!"
 fi
 
-if [[ "$oldHash" =~ ^(sha256|sha512)[:-] ]]; then
+if [[ "$oldHash" =~ ^(sha256|sha512|md5)[:-] ]]; then
```
Comment (jtojnar) [inline pkgs/common-updater/scripts/update-source-version:168]: https://github.com/NixOS/nixpkgs/issues/77238

Comment (nagisa) [reply pkgs/common-updater/scripts/update-source-version:168]: Aha, fair point. I can't do anything about the md5 hash coming from the API, but I guess I can just make the update script ignore the updates for packages that only provide an md5 hash and force manual updates for these.

Comment (nagisa) [reply pkgs/common-updater/scripts/update-source-version]: Fixed. I realized that although having `update-source-version` download the file to hash does not work, I can do the same just fine in the grafana's update script myself.

**Comment by nagisa** [comment]
Could I get a merge/review here?

File: pkgs/servers/monitoring/grafana/plugins/grafana-oncall-app/default.nix [hunk 3]
```diff
@@ -3,8 +3,8 @@
 grafanaPlugin {
   pname = "grafana-oncall-app";
   versionPrefix = "v";
-  version = "1.8.5";
-  zipHash = "sha256-HuZYHPTWm0EPKQbmapALK2j+PzM+J7gcWM9w8vU2yI0=";
+  version = "v1.9.26";
+  zipHash = "sha256-SMXjyU3OCHfUVPrSbgKkpXF/Hk6J8LpUg+kjJZNc/bE=";
```
Comment (Ma27) [inline pkgs/servers/monitoring/grafana/plugins/grafana-oncall-app/default.nix:7]: Fails with

```
grafana-oncall-app-vv1.9.26.zip> trying https://grafana.com/api/plugins/grafana-oncall-app/versions/vv1.9.26/download
grafana-oncall-app-vv1.9.26.zip>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
grafana-oncall-app-vv1.9.26.zip>                                  Dload  Upload   Total   Spent    Left  Speed
grafana-oncall-app-vv1.9.26.zip>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
grafana-oncall-app-vv1.9.26.zip> curl: (22) The requested URL returned error: 404
grafana-oncall-app-vv1.9.26.zip> error: cannot download grafana-oncall-app-vv1.9.26.zip from any mirror
```

I believe this is a bug in the update script?

**Review by Ma27 — COMMENTED** [review]
We usually have commits for each logical unit such as adding a new package. I think this should also be done for 4b09eba814744138dfea7920abccb02c5491a7f0.

File: pkgs/servers/monitoring/grafana/plugins/update-grafana-plugin.sh [hunk 4]
```diff
@@ -1,8 +1,37 @@
...
+        local urlPath="$(echo "$pkg" | jq -er .downloadUrl)"
+        local hash="$(nix-prefetch-url --type sha256 --name "$plugin_name" "https://grafana.com$urlPath")"
+    fi
+    hash=$(nix --extra-experimental-features nix-command hash to-sri --type sha256 "$hash")
+    # Set version number to a random number first to force update to happen
+    update-source-version $system "grafanaPlugins.${plugin_name}" $RANDOM "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
+    update-source-version $system "grafanaPlugins.${plugin_name}" "$latest_version" "$hash"
+}
+
+if echo "$api_response" | jq -e .packages.any; then
```
[the remark this answers is not in the document]
Comment (nagisa) [reply pkgs/servers/monitoring/grafana/plugins/update-grafana-plugin.sh:29]: I have chosen to drop the `-x` entirely for now.

File: pkgs/servers/monitoring/grafana/plugins/grafana-oncall-app/default.nix (hunk 3 above)
Comment (nagisa) [reply pkgs/servers/monitoring/grafana/plugins/grafana-oncall-app/default.nix:7]: Yeah, the common package thingy does weird stuff with `versionPrefix` and I'm not sure if its the responsibility of the updated script or `update-package-version` to deal with it.

Regardless it seems like in the time between my update and your review the upstream has yet again dropped the `v` prefix so I don't think there is any action necessary besides just removing `versionPrefix` from the package.

**Comment by nagisa** [comment]
@Ma27 I have addressed the review comments.

File: pkgs/servers/monitoring/grafana/plugins/update-grafana-plugin.sh [hunk 5]
```diff
@@ -1,8 +1,37 @@
...
+    if echo "$pkg" | jq -er .sha256 > /dev/null; then
+        local hash="$(echo "$pkg" | jq -er .sha256)"
+    else
+        # Some packages only have an md5 hash. Download the file for hash computation.
+        local urlPath="$(echo "$pkg" | jq -er .downloadUrl)"
+        local hash="$(nix-prefetch-url --type sha256 --name "$plugin_name" "https://grafana.com$urlPath")"
+    fi
+    hash=$(nix --extra-experimental-features nix-command hash to-sri --type sha256 "$hash")
+    # Set version number to a random number first to force update to happen
+    update-source-version $system "grafanaPlugins.${plugin_name}" $RANDOM "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
```
[the remark this answers is not in the document]
Comment (nagisa) [reply pkgs/servers/monitoring/grafana/plugins/update-grafana-plugin.sh:34]: Wrote a comment in the source code.

File: pkgs/servers/monitoring/grafana/plugins/update-grafana-plugin.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ma27) [reply pkgs/servers/monitoring/grafana/plugins/update-grafana-plugin.sh:38]: I don't mind the `-x`, but the `| jq` produces output, doesn't it? That's what I'd like to see suppressed.

[the remark this answers is not in the document]
Comment (nagisa) [reply pkgs/servers/monitoring/grafana/plugins/update-grafana-plugin.sh]: Alright, fixed both this and the output from curl too. Should be good to go now.

File: pkgs/servers/monitoring/grafana/plugins/update-grafana-plugin.sh [hunk 6]
```diff
@@ -1,8 +1,46 @@
...
+    # attempts to improve this script where I spent multiple hours investigating why the update
+    # script would refuse to update a hash that I intentionally malformed (in hopes to test the
+    # operation of this script.)
+    update-source-version $system "grafanaPlugins.${plugin_name}" $RANDOM "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
+    update-source-version $system "grafanaPlugins.${plugin_name}" "$latest_version" "$hash"
+}
+
+if echo "$api_response" | jq -e .packages.any > /dev/null; then
+    # the package contains an "any" package, so there should be only one zipHash.
+    update "any"
```
Comment (arianvp) [inline pkgs/servers/monitoring/grafana/plugins/update-grafana-plugin.sh:40]: IDK if this is the right approach.

For example https://github.com/NixOS/nixpkgs/pull/435531 has an `any` attribute now. But I looked at it and it just contains the binaries for all platforms. which makes the plugin kind of big.   But the update script fails to update the old hashes that were already in place :/

**Pull request merged**